### DISCUSSION TITLE
UI: Time and Date

### DISCUSSION THREAD

**Discussion by Sineos** [phoddie/node-red-mcu#95] [Q&A] [answered]
I'm generating a time / date string and have it displayed via a text-node

```
const d = new Date;
const langCode = 'de-DE'
const options = {
    timeZone: 'Europe/Berlin',
    timeStyle: 'medium',
    dateStyle: 'medium',
    hour12: false,
};

msg.payload = d.toLocaleString(langCode, options)

return msg;
```

It seems the formatting options are ignored in NR-MCU:
* NR-PC: `27.02.2023, 08:15:57`
* NR-MCU: `Mon Feb 27 2023 08:15:57`

**Comment by phoddie** [comment, marked as answer]
That's expected and valid because the `options` parameter to `Date.prototype.toLocaleString` is [optional](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString#syntax). The parameter is specified by ECMA-402 (e.g. `Intl`). XS implements the JavaScript language standard (ECMA-262), but not ECMA-402.

**Reply by Sineos** [reply]
Thanks for this clarification. 
Does this mean there is no way to influence how date and time string are formatted? Which, I have to admit, would be unfortunate, since I think on a MCU UI date and time is kind of "mandatory" but the very limited screen space would require a shorter representation.

**Reply by phoddie** [reply]
I agree: it would be wonderful if a microcontroller could offer everything that a computer does.

> Does this mean there is no way to influence how date and time string are formatted? 

The full ECMA-262 `Date` object is available which provides access to the information needed to format a date anyway you like from a `Function` node.

**Reply by Sineos** [reply]
Modifying the Date object is not that complicated, I'll get this done but maybe one beginner question:
What would be the best approach to handle the timezone on the MCU, especially with DST changes?

**Reply by phoddie** [reply]
> What would be the best approach to handle the timezone on the MCU, especially with DST changes?

The standard JavaScript date object provides a set of methods that operate in local time (`getDate`, `getHours`, `getMinutes`, etc.) and a parallel set of methods that operate in UTC time (`getUTCDate`, `getUTCHours`, `getUTCMinutes`, etc). The former respects the timezone and seasonal adjustments (DST).

Not all microcontrollers handle DST, though I believe the ESP-IDF does at this point.

**Reply by Sineos** [reply]
When the MCU is booted without connection to the PC / Debugger the time is set to UTC via the NTP. How do I tell the MCU, in which timezone it is and whether it needs to respect DST?
_Please_ do not tell me that I have to care for timezone and DST manually. 🙈

**Reply by phoddie** [reply]
> How do I tell the MCU, in which timezone it is and whether it needs to respect DST? Please do not tell me that I have to care for timezone and DST manually

You certainly [can set](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/base/base.md#time) them manually.

Since the xsbug connection automatically initializes the timezone and DST values, those values can be cached in preferences.  When the device runs without the debugger, the timezone and DST can be restored from preferences. I've committed a new [setup/timezone](https://github.com/phoddie/node-red-mcu/blob/main/setuptimezone.js) module that does that for microcontroller build targets.

**Reply by Sineos** [reply]
This sounds really good. Is my understanding correct:
1. I could use a function node to set `timezone` and `dst`, by importing the Time class via https://nodered.org/docs/user-guide/writing-functions#loading-additional-modules and then use the respective functions.
2. How would the MCU know when it is time to switch the DST? Would I need to compare manually against my local switching times, e.g. last Sunday in March at 01:00 UTC?

**Reply by phoddie** [reply]
> I could use a function node to set timezone and dst, by...

Correct.

> How would the MCU know when it is time to switch the DST?

Someone has to tell it. DST is unstable. If you know a good solution, please share.

**Reply by Sineos** [reply]
> Someone has to tell it. DST is unstable. If you know a good solution, please share.

Unfortunately I don't.
In general my experiences with timezone offsets and DST are already painful (at least for me) with PC-NR. For example:

### Sensor data and writing historic data
* PC-NR internally uses UTC, which is good
* If you keep your historic data in something like InfluxDB etc, you can simply pass the timestamp along with the measured value
* It would be complete mayhem to store data with local time stamps

### Reading and displaying historical data in PC-NR
* If you now read back historical data and want to display it in NR, you often need to manually create the the time-ranges. E.g. if I allow the Dashboard user to query the PV production for a certain date / date or time range, this needs to be converted from his local time entry to the correct UTC time.
* For displaying the data and depending how you display it, you might again have to correct the queried data to be in local time and this correction is also unstable due to DST
* Luckily PC-NR is partially time-zone aware due to
   * Using the clients browser's local time
   * As I just learned the differences between ECMA-402 and ECMA-262
   * Relative ease of using libraries like `moment` or `luxon`

### MCU-NR
Currently trying to understand how this is going to work in MCU-NR since the above points no longer work
* No client browser's local time
* No ECMA-402
* No nifty libraries to accommodate to my low-code life

**Reply by phoddie** [reply]
> It would be complete mayhem to store data with local time stamps 

Indeed. But, JavaScript Dates are natively UTC. So, no problem.

> No client browser's local time

I don't understand this point. Local time is available too.

Anyway.... If your goal is to do exactly the same things you do on the PC in exactly the same way on an MCU, you will be disappointed as you sound. MCUs can do quite a bit, but they will be a PC.

**Reply by Sineos** [reply]
Don't get me wrong, please. My experience with with MCUs is severely limited, so I just try to figure out, where and how I have to adapt. On a full fledged PC there is often "someone who is taking care about something, somehow", so I do not need to bother. On a MCU I need to learn about the differences. Many thanks for bearing with me so far. This is highly appreciated :+1

> I don't understand this point. Local time is available too.

If I understood our conversation correctly so far: Yes, true until DST kicks in spoils the fun. Would not be relevant to a MCU attached to the PC but relevant for a standalone MCU application. Meaning I need to do something in a function node to catch the DST switch times and adapt my local time. Correct?

**Reply by phoddie** [reply]
It looks like [with some work](https://github.com/G6EJD/ESP32-Time-Services-and-SETENV-variable), it is possible to configure the ESP-IDF to manage the DST offset. That's an adventure to explore.

**Reply by tve** [reply]
On a linux system `tail -1 /etc/localtime` will get you the desired DST config string, assuming your system has the time zone you're looking for...

**Reply by Sineos** [reply]
For the time being I solved it like this

<details>
<summary>function node</summary>

```js
// The end of DST in Europe is the last Sunday in March of the given year at 01:00 UTC
function getLastSundayInMarch() {
    const today = new Date();
    const year = today.getUTCFullYear();
    const lastDayOfMarch = new Date(Date.UTC(year, 2, 31, 1)); // March 31st at 01:00 UTC
    const dayOfWeek = lastDayOfMarch.getUTCDay();

    // If the last day of March is a Sunday, set the time to 01:00 UTC and return it.
    if (dayOfWeek === 0) {
        lastDayOfMarch.setUTCHours(1);
        return lastDayOfMarch;
    }

    // Otherwise, calculate the number of days to subtract from the last day of March to get to the last Sunday.
    const daysToSubtract = dayOfWeek;
    const lastSundayInMarch = new Date(Date.UTC(year, 2, 31 - daysToSubtract, 1)); // Last Sunday in March at 01:00 UTC
    return lastSundayInMarch;
}

// The begin of DST in Europe is the last Sunday in October of the given year at 01:00 UTC
function getLastSundayInOctober() {
    const today = new Date();
    const year = today.getUTCFullYear();
    const lastDayOfOctober = new Date(Date.UTC(year, 9, 31, 1)); // October 31st at 01:00 UTC
    const dayOfWeek = lastDayOfOctober.getUTCDay();

    // If the last day of October is a Sunday, set the time to 01:00 UTC and return it.
    if (dayOfWeek === 0) {
        lastDayOfOctober.setUTCHours(1);
        return lastDayOfOctober;
    }

    // Otherwise, calculate the number of days to subtract from the last day of October to get to the last Sunday.
    const daysToSubtract = dayOfWeek;
    const lastSundayInOctober = new Date(Date.UTC(year, 9, 31 - daysToSubtract, 1)); // Last Sunday in October at 01:00 UTC
    return lastSundayInOctober;
}

// Return a date / time string
function formatDate(date) {
    const day = String(date.getUTCDate()).padStart(2, '0');
    const month = String(date.getUTCMonth() + 1).padStart(2, '0');
    const year = String(date.getUTCFullYear()).slice(2);
    const hour = String(date.getUTCHours()).padStart(2, '0');
    const minute = String(date.getUTCMinutes()).padStart(2, '0');
    const second = String(date.getUTCSeconds()).padStart(2, '0');

    return `${day}.${month}.${year}, ${hour}:${minute}:${second}`;
}

// Correct the UTC time according to summer / winter DST offsets
function getDSTdate(date, dstOffsetSummer, dstOffsetWinter) {
    const lastSundayInMarch = getLastSundayInMarch();
    const lastSundayInOctober = getLastSundayInOctober();
    const dateTimestamp = date.getTime();

    if (dateTimestamp >= lastSundayInOctober.getTime() || dateTimestamp < lastSundayInMarch.getTime()) {
        date.setHours(date.getHours() + dstOffsetWinter);
    } else if (dateTimestamp >= lastSundayInMarch.getTime()) {
        date.setHours(date.getUTCHours() + dstOffsetSummer);
    }

    return formatDate(date);
}

msg.payload = getDSTdate(new Date(), 2, 1);

return msg;

```

</details>

While it seems to work, it is probably a very complicated way of doing it. So please do not laugh (at least not too loud) 😝
